pkg/store/bucket.go

```diff
@@ -382,6 +400,27 @@ func (s *BucketStore) numBlocks() int {
 	return len(s.blocks)
 }
 
+func (s *BucketStore) isBlockInMinMaxRange(ctx context.Context, id ulid.ULID) (bool, error) {
+	dir := filepath.Join(s.dir, id.String())
+
+	b := &bucketBlock{
+		logger: s.logger,
+		bucket: s.bucket,
+		id:     id,
+		dir:    dir,
+	}
+	if err := b.loadMeta(ctx, id); err != nil {
+		return false, err
+	}
+
+	// We check for blocks in configured minTime, maxTime range
+	if b.meta.MinTime < s.minTime.PrometheusTimestamp() || b.meta.MaxTime > s.maxTime.PrometheusTimestamp() {
```

Comment:
I think you want to compare against a blocks mintime, or maxtime, but not both.  The scenario I have in mind is something like this:

- You have a block that covers 01:00:00 to 03:00:00.
- You have a `thanos-store` process covering 00:00:00 to 02:00:00
- You have a second `thanos-store` process covering 02:00:01 to 04:00:00

I think with the logic above, *neither* thanos-store process will load the block because it's mintime is inside the first `thanos-store` process' range, but the maxtime is outside of its range.  For the second `thanos-store` you have the opposite situation.  Since neither of them own the *entire* block, neither of them will serve it.

Reply:
Yea I think essentially we need this:
```suggestion
	if b.meta.MaxTime < s.minTime.PrometheusTimestamp() || b.meta.MinTime > s.maxTime.PrometheusTimestamp() {
```

 So:
```
Store range:                         |min --------------max|
Blocks:                  | ------ A--------| | -------- B --------|
```

Thanks for spotting this @claytono 

Also can we have tests that would catch this?

Reply:
@bwplotka I think in that scenario, we will end up with the same block served by two stores.  For example:

```
Store A range: |min --- max|
Store B range              |min --- max|
Store C range                          |min --- max|
Blocks:        |----- A -----|----- B -----|----- C -----|
```

In this scenario, I think we want blocks A, B and C to be served by a single store.  With my previous PR, I picked minTime as the determining factor, so with that PR the assignments would be like this:

- Block A -> Store A 
- Block B -> Store A
- Block C -> Store B

I'm not sure there is a "right" answer for whether or not we should use minTime or maxTime for assigning blocks to stores, but I think in the interest of efficiency, if store time range are contiguous, then each block should only be owned by a single store.

Reply:
@bwplotka @claytono 

IMO it's fine in current version. i.e. users just need to be careful then setting right parameters.

So you set `|minTime - maxTime|` and store only will respond with blocks that fit that range. Making thanos store returning blocks outside the range is I think broken behavior, as it will just confuse people, why `| mnTime - maxTime|` isn't working.


I'll add docs for that

Reply:
BTW there are tests for current behaviour. If we decide to change them we can refactor those.

Reply:
I think it depends on what your use case is.   If the implementation requires that blocks must be entirely within the time range for the store, it may be impossible to pick time ranges where all blocks will be handled by a store, without having overlapping time ranges, which has other efficiency issues. 

Our primary desire for time based ranges is to be able to partition by time ranges, so the behavior as implemented now doesn't work well for us.  Ss the compactor generates larger and larger blocks, overlap between blocks with different labels makes it more and more likely that you're going to be serving very large blocks from multiple stores.  This means you'll be loading large blocks into memory in multiple places and wasting memory.  For us, memory usage of the store processes is the primary cost for Thanos.

Perhaps this should be configurable if different people want different behavior.